rclrs/src/node/service.rs

```diff
@@ -0,0 +1,168 @@
+use std::borrow::Borrow;
+use std::boxed::Box;
+use std::ffi::CString;
+use std::sync::Arc;
+
+use crate::error::{RclReturnCode, ToResult};
+use crate::Node;
+use crate::{rcl_bindings::*, RclrsError};
+
+use rosidl_runtime_rs::Message;
+
+use crate::node::publisher::MessageCow;
+
+use parking_lot::{Mutex, MutexGuard};
+
+// SAFETY: The functions accessing this type, including drop(), shouldn't care about the thread
+// they are running in. Therefore, this type can be safely sent to another thread.
+unsafe impl Send for rcl_service_t {}
+
+pub struct ServiceHandle {
+    handle: Mutex<rcl_service_t>,
+    node_handle: Arc<Mutex<rcl_node_t>>,
+}
+
+impl ServiceHandle {
+    pub fn lock(&self) -> MutexGuard<rcl_service_t> {
+        self.handle.lock()
+    }
+}
+
+impl Drop for ServiceHandle {
+    fn drop(&mut self) {
+        let handle = self.handle.get_mut();
+        let node_handle = &mut *self.node_handle.lock();
+        unsafe {
+            rcl_service_fini(handle as *mut _, node_handle as *mut _);
+        }
+    }
+}
+
+/// Trait to be implemented by concrete Service structs
+/// See [`Service<T>`] for an example
+pub trait ServiceBase: Send + Sync {
+    fn handle(&self) -> &ServiceHandle;
+    fn execute(&self) -> Result<(), RclrsError>;
+}
+
+/// Main class responsible for subscribing to topics and receiving data over IPC in ROS
+pub struct Service<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    pub handle: Arc<ServiceHandle>,
+    // The callback's lifetime should last as long as we need it to
+    pub callback:
+        Mutex<Box<dyn FnMut(&rmw_request_id_t, &T::Request, &mut T::Response) + 'static + Send>>,
+}
+
+impl<T> Service<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    pub fn new<F>(node: &Node, topic: &str, callback: F) -> Result<Self, RclrsError>
+    where
+        T: rosidl_runtime_rs::Service,
+        F: FnMut(&rmw_request_id_t, &T::Request, &mut T::Response) + 'static + Send,
```

Comment:
I understand this callback signature is meant to match the callback signature used in the original `create_service` definition used by `rclcpp` which has the benefit of being familiar for long-term users of `rclcpp`. However, the `rclcpp` API has come a long way since then and can now support async services with the [defer response](https://github.com/ros2/rclcpp/blob/86a9d5882ed31d1fa6197a1f5d4b388f167bd23e/rclcpp/include/rclcpp/any_service_callback.hpp#L214-L224) callback signatures.

This older fashioned API that uses `&mut T::Response` is unable to support non-blocking async services since it demands that a response be provided by the time the service callback exits. It also comes with the issue that `T::Response` must be getting implicitly default-initialized before it is passed in as a mutable reference. Implicit initialization is (rightly imo) seen as an anti-pattern in Rust.

I would recommend a callback signature that gives the benefits of supporting async services while still allowing ease of use:

```
Fn(&rmw_request_id_t, &T::Request, FnOnce(T::Response))
```

The `FnOnce` given to the callback will ensure
1. A service response is only provided once
2. The correct response is sent for the correct request

Changing to `Fn` instead of `FnMut` for the service callback trait should also help the signature to support multi-threaded executors for single services in the future.

Users that are okay with the service being sync can simply call the `FnOnce(T::Response)` before the function exits. Right now we unfortunately don't have an obvious way to force them to call it, but if the [function parameter attributes RFC](https://rust-lang.github.io/rfcs/2565-formal-function-parameter-attributes.html) makes it into the language then we'll be able to put `#[must_use]` on the response callback in the future. Meanwhile users that want an async service can pass the response callback off to some async function which can trigger the response callback when its work is finished.

Since I'm still not sure if `FnOnce` is usable here, we could alternatively have the response callback be something like

```
enum ResponseErr {
    AlreadyResponded,
}

FnMut(T::Response) -> Result<(), ResponseErr>
```

Inside of `FnMut` we would track whether a response has already been received by this callback and issue a `ResponseErr::AlreadyResponded` and ignore the input if it gets called a second time.

If any of this would be of interest, I'm happy to try prototyping it.

Reply:
Also if we want a callback signature that is explicitly sync and guarantees an immediate response, I would recommend making it

```
Fn(&rmw_request_id_t, &T::Request) -> T::Response
```

to move away from the implicit initialization problem.

Having both a sync and optionally async callback signature and allowing the user to choose is always nice.

Reply:
Good idea about avoiding implicit initialization, I'm rewriting the callback signature

Reply:
Done in 4607232